### ISSUE TITLE
Concept / spec

### ISSUE THREAD

**Issue by ryanblock** [metronome/metronome#1]
https://twitter.com/ryan/status/1157009541625573377
https://twitter.com/ryan/status/1157336684096155648

[tbd]

**Comment by erunion** [comment]
Ideas I had in no specific order:

- You should be able to designate a whole repository, or a directory within a repository, as something to sync. Anytime a file is changed within that directory/repo/whatever, that directory/repo/whatever would get synced across to the target replica repository.
- You should also be able to designate a target location for the sync to happen. So say you syncing something like a `legal/` folder in a `company/policies` repository that has Markdown, maybe you want that markdown to go into a `view/legal/` directory in the replica.
- All syncs should be treated as proposals and Metronome should generate a PR for approval similar to how @dependabot works.

**Comment by ryanblock** [comment]
Could imagine a TON of config UI for this - maybe we just start by reading a JSON config file from the master repo?

**Comment by erunion** [comment]
Yeah I wouldn't want to get bogged down in building a UI to get it going. Start with a simple JSON config to start and go from there.

**Comment by ryanblock** [comment]
Let's collect considerations here:

### IIRC, GH App installs are scoped to a given account/org
Example:
- User Bodega has personal repositories and is a member of org @dogrammers
- Bodega installs Metronome to her account
  - Metronome has access to her personal repos (e.g. `github.com/bodega/woof`)
  - However, Metronome cannot access @dogrammers repos
- Bodega gains admin privs to @dogrammers, yay! 🎉
  - Bodega can now create a new Metronome installation into @dogrammers
  - After doing so, the new Metronome installation can now access @dogrammers repos

(Pretty sure how that's how it works, will see.)

[more here?]

**Comment by ryanblock** [comment]
Example config ideas (in `json5` because comments); assuming user `bodega`.

Let's tear it to pieces!

#### Define `config.allRepos`
```json5
{
  allRepos: [
    'bodega/foo',   // Valid but redundant
    '@bodega/bar',  // Same
    'baz',          // bodega/baz
    'buz'           // bodega/buz
  ],
  files: {} // Use all files in this repo (except `metronome.json`), with default name + path
  // Also not defining files should be the same as an empty `files` object(?)
}
```

#### `config.allRepos === true`
```json5
{
  allRepos: true, // Attempt to use all repos in this org(?)
  files: {}
}
```

#### `config.files`, `config.files[file].repos`, `config.files[file].filename`
```json5
{
  // Define all scoped repos here
  allRepos: [
    'foo',
    'bar',
    'bodega' // Would attempt to use bodega/bodega
  ],
  // Define files from this repo to Ⓜ️etronome
  files: {
    'LICENSE': {}, // All repos, same filename
    '.gitignore': {
      // '/./gitignore' also valid; strip leading slashes (eg file.substr(-1) === path.sep ? file.substr(1) : file)
      repos: ['foo', 'bar']
    },
    'code_of_conduct.md': {
      repos: ['bar'],
      filename: '.github/code-of-conduct.md' // Transform folder + name
    }
  }
}
```

(Anything missing?)

**Comment by erunion** [comment]
* Why `allRepos` instead of just `repos`?
* This is the config file that would live within Metronome, or would this live within each repository replica repository?

**Comment by ryanblock** [comment]
> Why allRepos instead of just repos?

Mostly just for clarity / to prevent confusion with `config.files[file].repos`, but idc!

> This is the config file that would live within Metronome, or would this live within each repository replica repository?

(Good question – nomenclature check! `source` to refer to the originator repo of files, and `replica` to refer to the target/destination repo of files?)

I think there's just one of these config files in each source repo. BUT, we should probably support `.metronomeignore` to allow a replica to opt out of PRs.

wdyt?

**Comment by erunion** [comment]
Oh interesting, I was thinking that the replica repos would dictate what gets synced to them, not the primary dictating where its content goes.

**Comment by ryanblock** [comment]
Hmm, could do that! Let's tease that apart from a work to be done perspective:
### Config in source repo
**Update/delete**
- Change is made to file in source repo
- Event fires to Metronome
- Metronome then:
  - Looks at the change
  - Looks at the config
  - Builds a queue of jobs based on a single config
  - Fires off PRs to replica repos

**Create**
- User wants to add a new file to _n_ replica repos
- User adds it to source repo
- Event fires to Metronome
- Metronome then:
  - Looks at the change
  - Looks at the config
  - Builds a queue of jobs based on a single config
  - Fires off PRs to replica repos

### Config in replica repos
**Update/delete**
- Edit is made to file in source repo
- Event fires to Metronome
- Metronome polls all repos it has access to
- Metronome then:
  - Looks at the change
  - Looks at many replica configs
  - Builds a queue of jobs based on each config
  - Fires off PRs to replica repos

**Create**
- User wants to add a new file to _n_ replica repos
- User must then go all the replica repos (that don't subscribe to all files) and alter config to opt them into the new file
- User adds it to source repo
- Event fires to Metronome
- Metronome polls all repos it has access to
- Metronome then:
  - Looks at the change
  - Looks at many replica configs
  - Builds a queue of jobs based on each config
  - Fires off PRs to replica repos

Barring incorrect assumptions or any missing considerations, seems dramatically simpler for both users to configure and for us to build a central config. But I may be wrong!

**Comment by erunion** [comment]
I'm still not sure on having a source repo config, but from a basic flow standpoint that all looks okay to me.

I'm mostly apprehensive about the source repo stuff because what if you're setting up Metronome to monitor boilerplate repository that you don't own? Like if I want to replicate the [Contributor Covenant](https://github.com/ContributorCovenant/contributor_covenant) across all of my open source libraries.

**Comment by ryanblock** [comment]
Yooooooo that is a great point. Sounds like we should do both.

\*whispers* scope creeeeeeeep

**Comment by ryanblock** [comment]
Actually, thinking about this a bit more, not sure it would work for an MVP – certainly not easily.

In this example, you don't have CC admin privs, so when CC updates their repo, we should assume Metronome isn't installed. Therefore Metronome doesn't get an event from GitHub that a change occurred – thus, the only way to get those updates is to:
a) be an admin and add Metronome to every important repo out there, or
b) poll for updates (expensive and complex)

**Comment by erunion** [comment]
Oof that's a good point, I didn't think about having to poll for updates because we won't have hooks set up.

Punting non-owned repos to v2 I guess!
